x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,243 @@
+
+interface NodeStats {
+  totalEvents: number;
+  totalAlerts: number;
+}
+
+interface NodePagination {
+  nextChild?: string | null;
+  nextEvent?: string | null;
+  nextAncestor?: string | null;
+  nextAlert?: string | null;
+}
+
+interface Node {
+  id: string;
+  children: Node[];
+  events: ResolverEvent[];
+  alerts: ResolverEvent[];
+  lifecycle: ResolverEvent[];
+  parent?: Node | null;
+  pagination: NodePagination;
+  stats?: NodeStats;
+}
+
+type ExtractFunction = (event: ResolverEvent) => string;
+
+function createNode(id: string): Node {
+  return { id, children: [], pagination: {}, events: [], alerts: [], lifecycle: [] };
+}
+
+// This class aids in constructing a tree of process events. It works in the following way:
+//
+// 1. We construct a tree structure starting with the root node for the event we're requesting.
+// 2. We leverage the ability to pass hashes and arrays by reference to construct a fast cache of
+//    process identifiers that updates the tree structure as we push values into the cache.
+//
+// When we query a single level of results for child process events we have a flattened, sorted result
+// list that we need to add into a constructed tree. We also need to signal in an API response whether
+// or not there are more child processes events that we have not yet retrieved, and, if so, for what parent
+// process. So, at the end of our tree construction we have a relational layout of the events with no
+// pagination information for the given parent nodes. In order to actually construct both the tree and
+// insert the pagination information we basically do the following:
+//
+// 1. Using a terms aggregation query, we return an approximate roll-up of the number of child process
+//    "creation" events, this gives us an estimation of the number of associated children per parent
+// 2. We feed these child process creation event "unique identifiers" (basically a process.entity_id)
+//    into a second query to get the current state of the process via its "lifecycle" events.
+// 3. We construct the tree above with the "lifecycle" events.
+// 4. Using the terms query results, we mark each non-leaf node with the number of expected children, if our
+//    tree has less children than expected, we create a pagination cursor to indicate "we have a truncated set
+//    of values".
+// 5. We mark each leaf node (the last level of the tree we're constructing) with a "null" for the expected
+//    number of children to indicate "we have not yet attempted to get any children".
+//
+// Following this scheme, we use exactly 2 queries per level of children that we return--one for the pagination
+// and one for the lifecycle events of the processes. The downside to this is that we need to dynamically expand
+// the number of documents we can retrieve per level due to the exponential fanout of child processes,
+// what this means is that noisy neighbors for a given level may hide other child process events that occur later
```

Comment:
This is where a top hits aggregation would be useful.  When getting the children of multiple nodes simultaneously, we can use top hits to make sure we get at least N from each parent node instead of potentially getting a ton of children from a single parent node.  We don't need to support getting any page except page 1 when getting children from multiple nodes so the lack of search_after on top hits isn't an issue.  We can still return the pagination parameters for use with the other APIs that do support pagination.